js/optional-chaining-is-amazing-heres-why/optional-chaining-is-amazing-here-is-why.md

```diff
@@ -0,0 +1,154 @@
+```
+
+This magic solution has been achieved through **Optional Chaining (?.) Operator**.
+
+Now let's see how this Operator works with different data types in javascript. It only works with **Arrays**, **Objects** & **Functions**.
+
+**Optinal Chaining with Dynamic Properties**
+
+Optional Chaining works well with the dynamic properties with a different syntax. Let us see how:
+
+> //  The syntax for dynamic properties:
+  ?.[ ]
+
+```javascript
+const userdata = {
+    data : {
+        user: {
+            name: Ram Kumar,
+            socialMediaAccounts: ['twitter', 'linkedIn'],
+            primarySocialMedia: 'twitter',
+            socialLinks: {
+                twitter: 'https://twitter.com/mishraaSoumya',
+                linkedIn: 'https://www.linkedin.com/in/mishraa-soumya/'
+            }
+        }
+    }
+}
+
+const primarySocialMedia = data?.user?.primarySocialMedia;
+// the value of primarySocialMedia can be different or dynamic property.
+const socialMediaIUrl = data?.user?.socialLinks?.[primarySocialMedia];
+```
+
+Let me explain the above example how it works. So, In an app, a user can have multiple socialMediaAccounts, but there can be only 1 primary account, which can be different and dynamic. So, while pulling the url for primary social media account, a developer has to ensure that the dynamic value is a property in socialLinks object.
+
+The same syntax works with the **arrays** also, where you are not sure if the index will always be part of the array. Check the below example:
+
+```javascript
+// If the `usersArray` is `null` or `undefined`,
+// then `userName` gracefully evaluates to `undefined`.
+const userIndex = 13;
+const userName = usersArray?.[userIndex].name;
+```
+
+**Optional Chaining with Function Calls**
+
+Optional Chaining operator also works with Function Calls, but with an additional syntax form.
+
+> // Syntax for Function Calls
+?.( )
+
+```typescript
+interface InputProps = {
+    value: string;
+    onBlur?: Function;
+    onChange: Function;
+}
+
+const inputProps: InputProps = {
+    value,
+    onBlur,
+    onChange
+}
+
+const onBlurHandler = inputProps?.onBlur?.();
+```
+
+So, now you don't have to write multiple checks for your component props before actually calling the function.
+
+**Properties of Optional Chaining**
+
+The optional chaining operator has few properties: *short-circuiting*, *stacking*, and *optional deletion*. Let's go through these properties with an example.
+
+1. **Short-Circuiting**: It means the if LHS of the expression evaluates to null or undefined, then the RHS is not evaluated.
+
+> a?.[++x]
+
+In the above example, x is only incremented, if a is not null/undefined.
+
+2. **Stacking**: It means that more than 1 optional chaining operator can be used on a sequence of property accesses. Also, you can use different forms of the operator in the same sequence.
+
+```javascript
+const handler = db?.user?.[13]?.onClick?.();
```

Comment:
I'm a bit confused here. You mentioned that `[x]` syntax is used for dynamic evaluation. Here `13` is a const. 🤔

Reply:
@swapagarwal Through this example, I want to show that all the 3 forms of optional chaining operator can be used in 1 sequence. Also, I have updated the **const** to **let**

Reply:
Got it, my bad! Whatever is within square brackets gets evaluated.